### ISSUE TITLE
3_train.py

### ISSUE THREAD

**Issue by SebliLo** [BMIRDS/deepslide#46]
It seems the network is not learning, just 0.5 for both training and validation accuracies. 
The same dataset with other network provides .85 accuracy. 
What would be your suggestion? 

###############     CONFIGURATION     ###############
all_wsi:	all_wsi
val_wsi_per_class:	20
test_wsi_per_class:	30
keep_orig_copy:	True
num_workers:	8
patch_size:	224
wsi_train:	wsi_train
wsi_val:	wsi_val
wsi_test:	wsi_test
labels_train:	labels_train.csv
labels_val:	labels_val.csv
labels_test:	labels_test.csv
train_folder:	train_folder
patches_eval_train:	patches_eval_train
patches_eval_val:	patches_eval_val
patches_eval_test:	patches_eval_test
num_train_per_class:	8000
type_histopath:	True
purple_threshold:	100
purple_scale_size:	15
slide_overlap:	3
gen_val_patches_overlap_factor:	1.5
image_ext:	jpg
by_folder:	True
color_jitter_brightness:	0.5
color_jitter_contrast:	0.5
color_jitter_saturation:	0.5
color_jitter_hue:	0.2
num_epochs:	20
num_layers:	18
learning_rate:	0.001
batch_size:	16
weight_decay:	0.0001
learning_rate_decay:	0.85
resume_checkpoint:	False
save_interval:	1
checkpoints_folder:	checkpoints
checkpoint_file:	xyz.pt
pretrain:	False
log_folder:	logs
auto_select:	True
preds_train:	preds_train
preds_val:	preds_val
preds_test:	preds_test
inference_train:	inference_train
inference_val:	inference_val
inference_test:	inference_test
vis_train:	vis_train
vis_val:	vis_val
vis_test:	vis_test
device:	cpu
classes:	['benign', 'tumor']
num_classes:	2
train_patches:	train_folder/train
val_patches:	train_folder/val
path_mean:	[3.2604160257049264e-12, 1.7743442033415775e+28, 2.053507277676774e-19]
path_std:	[nan, nan, nan]
resume_checkpoint_path:	checkpoints/xyz.pt
log_csv:	logs/log_262022_213727.csv
eval_model:	checkpoints/xyz.pt
threshold_search:	(0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9)
colors:	('red', 'white', 'blue', 'green', 'purple', 'orange', 'black', 'pink', 'yellow')

#####################################################

+++++ Running 3_train.py +++++
2 classes: ['benign', 'tumor']
num train images 1056
num val images 48
CUDA is_available: False
train_folder: train_folder
num_epochs: 20
num_layers: 18
learning_rate: 0.001
batch_size: 16
weight_decay: 0.0001
learning_rate_decay: 0.85
resume_checkpoint: False
resume_checkpoint_path (only if resume_checkpoint is true): checkpoints/xyz.pt
save_interval: 1
output in checkpoints_folder: checkpoints
pretrain: False
log_csv: logs/log_262022_213727.csv

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 0 with lr 0.000850000000000: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 1 with lr 0.000722500000000: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 2 with lr 0.000614125000000: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 3 with lr 0.000522006250000: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 4 with lr 0.000443705312500: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 5 with lr 0.000377149515625: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 6 with lr 0.000320577088281: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 7 with lr 0.000272490525039: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 8 with lr 0.000231616946283: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 9 with lr 0.000196874404341: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 10 with lr 0.000167343243690: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 11 with lr 0.000142241757136: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 12 with lr 0.000120905493566: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 13 with lr 0.000102769669531: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 14 with lr 0.000087354219101: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 15 with lr 0.000074251086236: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 16 with lr 0.000063113423301: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 17 with lr 0.000053646409806: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 18 with lr 0.000045599448335: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 19 with lr 0.000038759531085: t_loss: nan t_acc: 0.5000 v_loss: nan v_acc: 0.5000

training complete in 53.00 minutes
+++++ Finished running 3_train.py +++++

**Comment by JosephDiPalma** [comment]
There is something wrong with your input data. The values of `path_mean` and `path_std` are incorrect and are causing this problem.

**Comment by SebliLo** [comment]
@JosephDiPalma  thank you for the input!
The value of path_mean and path_std are slide or patch level?  How I fix it?
The input data seems working on another neural network, I tried outside of this.

**Comment by JosephDiPalma** [comment]
@SebliLo Do you have access to the per-channel mean and standard deviation for your current dataset? If so, I can show you how to modify our code to use those values instead of computing them again.

**Comment by SebliLo** [comment]
@JosephDiPalma  No, I don't. 
The issue occurred because of the number of patches is low, about 1000 each for 2 classes? Have you tried with such a small dataset?

**Comment by SebliLo** [comment]
This is how I normalize the pixel values. 

images = np.array(images)
images = images.astype('float32') / 255.0 
train_mean = np.mean(images, axis=0)
train_std = np.std(images, axis=0)
images = (images - train_mean) / train_std
images, labels = shuffle(images, labels, random_state=1)

**Comment by JosephDiPalma** [comment]
Change the lines 368-369 in `config.py` to the following:
```
path_mean = [0.8637807, 0.7897088, 0.8675098]
path_std = [0.06827197, 0.10325973, 0.09146392]
```
Now the code will use these values for mean and standard deviation.

**Comment by SebliLo** [comment]
Thank you for the suggestion. I updated the path_mean and path_std.  Do you see any issue? 

+++++ Running 3_train.py +++++
2 classes: ['benign', 'tumor']
num train images 1056
num val images 48
CUDA is_available: False
train_folder: train_folder
num_epochs: 20
num_layers: 18
learning_rate: 0.001
batch_size: 16
weight_decay: 0.0001
learning_rate_decay: 0.85
resume_checkpoint: False
resume_checkpoint_path (only if resume_checkpoint is true): checkpoints/xyz.pt
save_interval: 1
output in checkpoints_folder: checkpoints
pretrain: False
log_csv: logs/log_2162022_211913.csv


Predicted  benign  tumor
Actual                  
benign       0.63   0.37
tumor        0.42   0.58
Predicted  benign  tumor
Actual                  
benign       0.55   0.45
tumor        0.05   0.95
Epoch 0 with lr 0.000850000000000: t_loss: 0.7437 t_acc: 0.6053 v_loss: 0.6076 v_acc: 0.7500

Predicted  benign  tumor
Actual                  
benign       0.66   0.34
tumor        0.36   0.64
Predicted  benign  tumor
Actual                  
benign       0.00   1.00
tumor        0.00   1.00
Epoch 1 with lr 0.000722500000000: t_loss: 0.6444 t_acc: 0.6499 v_loss: 0.7701 v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       0.75   0.25
tumor        0.36   0.64
Predicted  benign  tumor
Actual                  
benign       0.90   0.10
tumor        0.00   1.00
Epoch 2 with lr 0.000614125000000: t_loss: 0.5944 t_acc: 0.6964 v_loss: 0.3550 v_acc: 0.9500

Predicted  benign  tumor
Actual                  
benign       0.73   0.27
tumor        0.31   0.69
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        0.95   0.05
Epoch 3 with lr 0.000522006250000: t_loss: 0.5856 t_acc: 0.7059 v_loss: 0.8213 v_acc: 0.5250

Predicted  benign  tumor
Actual                  
benign       0.69   0.31
tumor        0.30   0.70
Predicted  benign  tumor
Actual                  
benign       0.05   0.95
tumor        0.00   1.00
Epoch 4 with lr 0.000443705312500: t_loss: 0.5846 t_acc: 0.6917 v_loss: 0.6550 v_acc: 0.5250

Predicted  benign  tumor
Actual                  
benign       0.70   0.30
tumor        0.24   0.76
Predicted  benign  tumor
Actual                  
benign       0.20   0.80
tumor        0.00   1.00
Epoch 5 with lr 0.000377149515625: t_loss: 0.5469 t_acc: 0.7334 v_loss: 0.5639 v_acc: 0.6000

Predicted  benign  tumor
Actual                  
benign       0.72   0.28
tumor        0.27   0.73
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        0.45   0.55
Epoch 6 with lr 0.000320577088281: t_loss: 0.5354 t_acc: 0.7239 v_loss: 0.5004 v_acc: 0.7750

Predicted  benign  tumor
Actual                  
benign       0.76   0.24
tumor        0.28   0.72
Predicted  benign  tumor
Actual                  
benign       0.00   1.00
tumor        0.00   1.00
Epoch 7 with lr 0.000272490525039: t_loss: 0.5096 t_acc: 0.7419 v_loss: 1.7465 v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       0.82   0.18
tumor        0.26   0.74
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        1.00   0.00
Epoch 8 with lr 0.000231616946283: t_loss: 0.4822 t_acc: 0.7780 v_loss: 0.7307 v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       0.83   0.17
tumor        0.25   0.75
Predicted  benign  tumor
Actual                  
benign       0.90   0.10
tumor        0.00   1.00
Epoch 9 with lr 0.000196874404341: t_loss: 0.4225 t_acc: 0.7913 v_loss: 0.1214 v_acc: 0.9500

Predicted  benign  tumor
Actual                  
benign       0.85   0.15
tumor        0.22   0.78
Predicted  benign  tumor
Actual                  
benign       0.00   1.00
tumor        0.00   1.00
Epoch 10 with lr 0.000167343243690: t_loss: 0.4045 t_acc: 0.8150 v_loss: 3.8903 v_acc: 0.5000

Predicted  benign  tumor
Actual                  
benign       0.89   0.11
tumor        0.22   0.78
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        0.00   1.00
Epoch 11 with lr 0.000142241757136: t_loss: 0.3755 t_acc: 0.8359 v_loss: 0.0545 v_acc: 1.0000

Predicted  benign  tumor
Actual                  
benign       0.83   0.17
tumor        0.18   0.82
Predicted  benign  tumor
Actual                  
benign       0.70   0.30
tumor        0.00   1.00
Epoch 12 with lr 0.000120905493566: t_loss: 0.3815 t_acc: 0.8254 v_loss: 0.3024 v_acc: 0.8500

Predicted  benign  tumor
Actual                  
benign       0.89   0.11
tumor        0.17   0.83
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        0.10   0.90
Epoch 13 with lr 0.000102769669531: t_loss: 0.3333 t_acc: 0.8596 v_loss: 0.1000 v_acc: 0.9500

Predicted  benign  tumor
Actual                  
benign       0.87   0.13
tumor        0.17   0.83
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        0.00   1.00
Epoch 14 with lr 0.000087354219101: t_loss: 0.3561 t_acc: 0.8491 v_loss: 0.0508 v_acc: 1.0000

Predicted  benign  tumor
Actual                  
benign       0.89   0.11
tumor        0.17   0.83
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        0.00   1.00
Epoch 15 with lr 0.000074251086236: t_loss: 0.3082 t_acc: 0.8596 v_loss: 0.0502 v_acc: 1.0000

Predicted  benign  tumor
Actual                  
benign       0.88   0.12
tumor        0.17   0.83
Predicted  benign  tumor
Actual                  
benign       0.55   0.45
tumor        0.00   1.00
Epoch 16 with lr 0.000063113423301: t_loss: 0.3240 t_acc: 0.8539 v_loss: 0.4242 v_acc: 0.7750

Predicted  benign  tumor
Actual                  
benign       0.91   0.09
tumor        0.17   0.83
Predicted  benign  tumor
Actual                  
benign       0.85   0.15
tumor        0.00   1.00
Epoch 17 with lr 0.000053646409806: t_loss: 0.3184 t_acc: 0.8700 v_loss: 0.2135 v_acc: 0.9250

Predicted  benign  tumor
Actual                  
benign       0.90   0.10
tumor        0.15   0.85
Predicted  benign  tumor
Actual                  
benign       0.90   0.10
tumor        0.00   1.00
Epoch 18 with lr 0.000045599448335: t_loss: 0.2908 t_acc: 0.8757 v_loss: 0.1171 v_acc: 0.9500

Predicted  benign  tumor
Actual                  
benign       0.91   0.09
tumor        0.12   0.88
Predicted  benign  tumor
Actual                  
benign       1.00   0.00
tumor        0.20   0.80
Epoch 19 with lr 0.000038759531085: t_loss: 0.2479 t_acc: 0.8985 v_loss: 0.1668 v_acc: 0.9000


training complete in 41.00 minutes
+++++ Finished running 3_train.py +++++

**Comment by JosephDiPalma** [comment]
The training looks fine now.

**Comment by SebliLo** [comment]
Thank you for your help.
